pkg/scheduler/controller.go

```diff
@@ -0,0 +1,249 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	kcpclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/cluster"
+	schedulingv1alpha1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/scheduling/v1alpha1"
+	workloadv1alpha1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/workload/v1alpha1"
+	schedulingv1alpha1listers "github.com/kcp-dev/kcp/pkg/client/listers/scheduling/v1alpha1"
+	workloadv1alpha1listers "github.com/kcp-dev/kcp/pkg/client/listers/workload/v1alpha1"
+
+	edgeclientset "github.com/kcp-dev/edge-mc/pkg/client/clientset/versioned/cluster"
+	edgev1alpha1informers "github.com/kcp-dev/edge-mc/pkg/client/informers/externalversions/edge/v1alpha1"
+	edgev1alpha1listers "github.com/kcp-dev/edge-mc/pkg/client/listers/edge/v1alpha1"
+	"github.com/kcp-dev/edge-mc/pkg/indexers"
+)
+
+const (
+	ControllerName = "edge-scheduler"
+)
+
+type triggeringKind string
+
+const (
+	triggeringKindEdgePlacement triggeringKind = "EdgePlacement"
+	triggeringKindLocation      triggeringKind = "Location"
+	triggeringKindSyncTarget    triggeringKind = "SyncTarget"
+)
+
+type queueItem struct {
+	triggeringKind triggeringKind
+	key            string
+}
+
+type controller struct {
+	context context.Context
+	queue   workqueue.RateLimitingInterface
+
+	kcpClusterClient  kcpclientset.ClusterInterface
+	edgeClusterClient edgeclientset.ClusterInterface
+
+	edgePlacementLister  edgev1alpha1listers.EdgePlacementClusterLister
+	edgePlacementIndexer cache.Indexer
+
+	locationLister  schedulingv1alpha1listers.LocationClusterLister
+	locationIndexer cache.Indexer
+
+	synctargetLister  workloadv1alpha1listers.SyncTargetClusterLister
+	synctargetIndexer cache.Indexer
+}
+
+func NewController(
+	context context.Context,
+	kcpClusterClient kcpclientset.ClusterInterface,
+	edgeClusterClient edgeclientset.ClusterInterface,
+	edgePlacementAccess edgev1alpha1informers.EdgePlacementClusterInformer,
+	locationAccess schedulingv1alpha1informers.LocationClusterInformer,
+	syncTargetAccess workloadv1alpha1informers.SyncTargetClusterInformer,
+) (*controller, error) {
+	context = klog.NewContext(context, klog.Background().WithValues("controller", ControllerName))
```

Comment:
What do you think of using `klog.FromContext(context)` instead of `klog.Background()` here?

Reply:
That makes sense, I tweaked it by https://github.com/kcp-dev/edge-mc/pull/206/commits/5a5cd2a9e9c74a3fd56d2120f37985877aedd50b.